### PR TITLE
Emphasize randomness importance for the blinding factor

### LINKED ISSUE #55: Emphasize randomness importance
From Jeff Burdges on the list:

> It’s critically important the blinding factor r be a uniformly random 
> integer mod n, which I think deserves more emphasis than you give.  
> There is an easy deanonymization attack if r were say generated a 
> random integer mod 2^{floor(log2 n)}.  You hould emphasize that 
> random_integer should be instantiated with a CSPRNG and rejection 
> sampling, maybe even specify the rejection sampling algorithm starting 
> with shake or chacha.

### PR DESCRIPTION
**Pull request by jedisct1**
- Add yet another emphasis on the fact that `r` must be chosen from a uniform distribution
- Suggest a way to implement `random_integer_uniform`, by generating and reducing a larger integer
- Update the Python code to do exactly that instead of rejection sampling (nothing wrong with rejection sampling, but the reduction method is shorter to document)

Fixes #55

### CODE REVIEW

File: draft-irtf-cfrg-rsa-blind-signatures.md [hunk 1]
```diff
@@ -245,6 +245,16 @@ Steps:
 11. output blinded_msg, inv
 ~~~
 
+The blinding factor r must be randomly chosen from a uniform distribution.
+
+In order to do so, the random_integer_uniform function can be implemented by obtaining L uniform bytes, where
+
+~~~
+L = ceil((ceil(log2(n - m)) + k) / 8)
```
Comment (FredericJacobs) [inline draft-irtf-cfrg-rsa-blind-signatures.md:253]: Does this heuristic come from somewhere we can reference? If not, we might want to add a line or two about how this was defined.

Comment (chris-wood) [reply draft-irtf-cfrg-rsa-blind-signatures.md:253]: It was included in [FIPS-186](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf), though not in this general form. We could just cite that?

**Review by chris-wood — APPROVED** [review]

**Comment by chris-wood** [comment]
Also, @tlepoint recently did some analysis [in a different context](https://github.com/abetterinternet/prio-documents/issues/14) comparing this approach with rejection sampling, and found the latter was more performant. Do we want to comment on this at all? It is an implementation-specific decision, after all.

**Comment by jedisct1** [comment]
Rejection sampling is almost always faster than a reduction, even though in the context of the our `Blind` function, the difference is negligible.

"how to generate unbiased random numbers" feels a little bit out of the scope of this document, so the intent was to keep that part as short as possible; the reduction method was a better candidate for this.

To be honest, documenting any of these methods may do more harm than good. Implementations should use what their cryptographic library already provides. Reimplementing this is rarely necessary and can only introduce new issues.

**Comment by chris-wood** [comment]
> To be honest, documenting any of these methods may do more harm than good. Implementations should use what their cryptographic library already provides. Reimplementing this is rarely necessary and can only introduce new issues.

You may be right. Maybe we can just remove the implementation entirely and say something like "this is typically done via rejection sampling or flooding"?

**Comment by jedisct1** [comment]
> You may be right. Maybe we can just remove the implementation entirely and say something like "this is typically done via rejection sampling or flooding"?

How about just "this is typically done via rejection sampling"? This doesn't need clarification, and the method is well understood and widely documented.

**Comment by chris-wood** [comment]
Works for me!

**Review by chris-wood — APPROVED** [review]

File: poc/rsabssa.py [hunk 2]
```diff
@@ -57,7 +58,10 @@ def RSAVP1(public_key: RsaKey, r: int) -> int:
     return pow(r, e, n)
 
 
-def random_integer_uniform(min: int, max: int) -> int:
+def random_integer_uniform(m: int, n: int) -> int:
+    # Implement rejection sampling.
+    # This is for reference only; most cryptographic libraries include
+    # functions to generate random integers from a uniform distribution.
     range = max - min
```
Comment (tlepoint) [inline poc/rsabssa.py:65]: This will not work anymore, since you change the names of the parameters

File: poc/rsabssa.py [hunk 3]
```diff
@@ -57,7 +58,10 @@ def RSAVP1(public_key: RsaKey, r: int) -> int:
...
     rangeBits = size(range - 1)
```
Comment (tlepoint) [inline poc/rsabssa.py:66]: You may want to indicate in the comment which of the intervals `[m, n)` and `[m, n]` this function samples from.

File: poc/rsabssa.py (hunk 2 above)
Comment (jedisct1) [reply poc/rsabssa.py:65]: Good catch, thank you!

File: poc/rsabssa.py (hunk 3 above)
Comment (jedisct1) [reply poc/rsabssa.py:66]: Added, thank you!

**Pull request merged**